### PR TITLE
Run compatibility check based on PR base branch

### LINKED ISSUE #9371: [BUG] Check compatibility doesn't run on the code in the pull request
**Describe the bug**
When the Check Compatibility workflow is run, it uses `pull_request_target` [1], this allows the check to have write access to the repository for updating comments; however, when built it isn't using the code from the pull request itself, so the check is always running against what is in main or 2.x at the time the pull request is updated.

**To Reproduce**
1. Go to a recent pull request, like https://github.com/opensearch-project/OpenSearch/pull/9366
2. Look for the Compatibility status comment
3. See the `compatible with change XXXXX` part of the message
4. Check which commit that change is associated with
5. Notice the change is from the `main` branch

**Expected behavior**
The commit should be from the pull request.

**Additional context**
- [pull_request_target](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target)

### PR DESCRIPTION
**Pull request by gaiksaya**
### Description
Run compatibility check based on PR base branch rather than main branch. See [pull_request_target](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) and [pull_request](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request)

### Related Issues
Resolves #9371 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by peternied — CHANGES_REQUESTED** [review]
> You can use and read encrypted secrets in a workflow file if you have access to edit the file.

https://docs.github.com/en/actions/security-guides/encrypted-secrets#accessing-your-secrets

When triggered by `pull_request` write permissions are not allowed so `app_id` and other secrets won't be resolved

**Comment by peternied** [comment]
@gaiksaya Any progress here?

**Comment by gaiksaya** [comment]
> @gaiksaya Any progress here?

Sorry for the delay. Updated the PR with the changes. Here is sample comment: https://github.com/gaiksaya/OpenSearch/pull/8

**Review by peternied — APPROVED** [review]
Nice!

**Review by kotwanikunal — APPROVED** [review]

File: .github/workflows/check-compatibility.yml [hunk 1]
```diff
@@ -7,9 +7,14 @@ on:
...
   build:
     if: github.repository == 'opensearch-project/OpenSearch'
+    permissions:
+      contents: read # to fetch code (actions/checkout)
+      pull-requests: write # to create or update comment (peter-evans/create-or-update-comment)
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
+        with:
+          ref: ${{ github.event.pull_request.head.sha }}
```
Comment (peternied) [inline .github/workflows/check-compatibility.yml:17]: This changes the environment of the GitHub action to get code from a pull request and then run `./gradlew checkCompatibility` with write permissions on pull requests.  This seems like it opens up a security vulnerability in the GitHub action.

I think we need a separate workflow that is triggered on the pull request that doesn't have access to the write permissions, we could have it output its results and then pull it in like how the gradlew-check.out is used

Comment (gaiksaya) [reply .github/workflows/check-compatibility.yml:17]: I think that was gradle check as it as well here https://github.com/opensearch-project/OpenSearch/blob/main/.github/workflows/gradle-check.yml#L16-L20

Comment (gaiksaya) [reply .github/workflows/check-compatibility.yml:16]: I added another job with write permission to only comment on the PR. Please see the latest commit for the changes. Thank you!

**Review by peternied — APPROVED** [review]

**Pull request merged**